### PR TITLE
Fix allowed user types setting problem

### LINKED ISSUE #25951: HTTP 500 while creating users
### Description
I tried to create a new user, user visibility was defaulted to private, but after clicking Create user - error 500.
Should it be a 500 error? May be a more user friendly message?
But I set user visibility to "Private", and Public users are indeed disallowed. And there was no "Public" item in visibility dropdown at all.

It looks like UI inconsistency between default dropdown selection and data passed to backend.
If I re-select Private in dropdown, user created without error.

image

### Gitea Version

1.20

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

win service

### Database

None

### PR DESCRIPTION
**Pull request by lunny**
Fix #25951

### CODE REVIEW

File: modules/setting/service.go [hunk 1]
```diff
@@ -188,15 +188,25 @@ func loadServiceFrom(rootCfg ConfigProvider) {
...
 		Service.AllowedUserVisibilityModesSlice = []bool{false, false, false}
 		for _, sMode := range Service.AllowedUserVisibilityModes {
 			Service.AllowedUserVisibilityModesSlice[structs.VisibilityModes[sMode]] = true
 		}
+	} else {
+		Service.AllowedUserVisibilityModes = []string{"public", "limited", "private"}
+	}
+	Service.DefaultUserVisibility = sec.Key("DEFAULT_USER_VISIBILITY").String()
+	if Service.DefaultUserVisibility == "" {
+		Service.DefaultUserVisibility = Service.AllowedUserVisibilityModes[0]
```
Comment (delvh) [inline modules/setting/service.go:212]: Can still error if `Service.AllowedUserVisibilityModes = []`.

Comment (delvh) [reply modules/setting/service.go:212]: Also below.
We certainly need a check before this line that the slice is not empty.

Comment (lunny) [reply modules/setting/service.go:212]: > Can still error if `Service.AllowedUserVisibilityModes = []`.

I think I have test case for that?

File: modules/setting/service.go [hunk 2]
```diff
@@ -188,15 +188,31 @@ func loadServiceFrom(rootCfg ConfigProvider) {
 	Service.EnableUserHeatmap = sec.Key("ENABLE_USER_HEATMAP").MustBool(true)
 	Service.AutoWatchNewRepos = sec.Key("AUTO_WATCH_NEW_REPOS").MustBool(true)
 	Service.AutoWatchOnChanges = sec.Key("AUTO_WATCH_ON_CHANGES").MustBool(false)
-	Service.DefaultUserVisibility = sec.Key("DEFAULT_USER_VISIBILITY").In("public", structs.ExtractKeysFromMapString(structs.VisibilityModes))
-	Service.DefaultUserVisibilityMode = structs.VisibilityModes[Service.DefaultUserVisibility]
-	Service.AllowedUserVisibilityModes = sec.Key("ALLOWED_USER_VISIBILITY_MODES").Strings(",")
-	if len(Service.AllowedUserVisibilityModes) != 0 {
+	modes := sec.Key("ALLOWED_USER_VISIBILITY_MODES").Strings(",")
+	if len(modes) != 0 {
```
Comment (wxiaoguang) [inline modules/setting/service.go:192]: It needs to reset the AllowedUserVisibilityModes

File: modules/setting/service.go [hunk 3]
```diff
@@ -188,15 +188,31 @@ func loadServiceFrom(rootCfg ConfigProvider) {
...
+				Service.AllowedUserVisibilityModes = append(Service.AllowedUserVisibilityModes, sMode)
+				Service.AllowedUserVisibilityModesSlice[tp] = true
+			} else {
+				log.Warn("ALLOWED_USER_VISIBILITY_MODES %s is unsupported", sMode)
+			}
 		}
 	}
+
+	if len(Service.AllowedUserVisibilityModes) == 0 {
+		Service.AllowedUserVisibilityModes = []string{"public", "limited", "private"}
```
Comment (wxiaoguang) [inline modules/setting/service.go:206]: I think it's better to set `AllowedUserVisibilityModesSlice = true, true, true` explicitly here.

**Review by wxiaoguang — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Pull request merged**